File: cascadingFailuresTaskMigration/src/main/java/evaluation/Evalution.java
package evaluation;

import input.Group;
import input.MigrationRecord;
import input.Robot;
import input.Task;
import main.Function;
import org.jgrapht.alg.interfaces.ShortestPathAlgorithm;
import org.jgrapht.graph.DefaultWeightedEdge;

import java.util.HashMap;
import java.util.List;

public class Evalution {
    HashMap<Integer, Robot> idToRobots;
    HashMap<Integer, Group> idToGroups;
    //计算迁移成本
    Function function;

    public Evalution(HashMap<Integer, Robot> idToRobots, HashMap<Integer, Group> idToGroups) {
        this.idToRobots = idToRobots;
        this.idToGroups = idToGroups;
        function = new Function(idToRobots, idToGroups);
    }

    public Double calculateMigrationCost(ShortestPathAlgorithm<Integer, DefaultWeightedEdge> shortestPath, List<MigrationRecord> migrationRecords) {
        Double sum = 0.0;
        //计算总的迁移成本
        for (MigrationRecord migrationRecord : migrationRecords) {
            double pathWeight = shortestPath.getPathWeight(migrationRecord.getFrom(), migrationRecord.getTo());
            sum += pathWeight;
        }


        return sum;
    }

    //计算完成任务的成本
    public Double calculateExecuteTasksCost(List<Robot> robots) {
        Double sum = 0.0;
        //Function function = new Function(idToRobots, idToGroups);
        for (Robot robot : robots) {
            List<Task> tasksList = robot.getTasksList();
            if (tasksList != null) {
                for (Task task : tasksList) {
                    sum += task.getSize() / robot.getCapacity();
                }
            }
        }
        return sum;
    }

    //计算平均存活率
    public Double calculateMeanSurvivalRate(List<Robot> robots) {
        Double sum = 0.0;
        int count = 0;
        for (Robot robot : robots) {
            if (robot.getFaultA()!=1) {
                count++;
            }
            Double survivalRate = (1-robot.getFaultA())*(1-robot.getFaultO());
            sum += survivalRate;
        }
        return sum / count;
    }


}

File: FAO_case_study/algorithms/evaluation/evalution.py
from typing import List, Dict
from input.robot import Robot
from input.group import Group
from input.migration_record import MigrationRecord
from main.function import Function


class Evalution:
    def __init__(self, id_to_robots: Dict[int, Robot], id_to_groups: Dict[int, Group]):
        self.id_to_robots = id_to_robots
        self.id_to_groups = id_to_groups
        self.function = Function(id_to_robots, id_to_groups)

    def calculate_migration_cost(self, shortest_path_dict: Dict, migration_records: List[MigrationRecord]) -> float:
        """Calculate total migration cost."""
        total = 0.0
        for record in migration_records:
            path_weight = shortest_path_dict.get((record.from_robot, record.to_robot), 0.0)
            total += path_weight
        return total

    def calculate_execute_tasks_cost(self, robots: List[Robot]) -> float:
        """Calculate total task execution cost."""
        total = 0.0
        for robot in robots:
            tasks_list = robot.tasks_list
            if tasks_list is not None:
                for task in tasks_list:
                    total += task.size / robot.capacity
        return total

    def calculate_mean_survival_rate(self, robots: List[Robot]) -> float:
        """Calculate mean survival rate."""
        total = 0.0
        count = 0
        for robot in robots:
            if robot.fault_a != 1:
                count += 1
            survival_rate = (1 - robot.fault_a) * (1 - robot.fault_o)
            total += survival_rate
        return total / count if count > 0 else 0.0
